File: YYRefresh/YYRefresh/YYRefreshView.m
//
//  YYRefreshView.m
//  ObjcSum
//
//  Created by sihuan on 2016/6/21.
//  Copyright © 2016年 sihuan. All rights reserved.
//

#import "YYRefreshView.h"
#import "YYRefreshConst.h"

#pragma mark - Const
// 文字颜色
#define YYRefreshLabelTextColor [UIColor colorWithWhite:0.400 alpha:1.000]
// 字体大小
#define YYRefreshLabelFont [UIFont boldSystemFontOfSize:14]

@interface YYRefreshView ()

@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UIActivityIndicatorView *refreshingView;
@property (nonatomic, strong) UILabel *textLabel;
@property (nonatomic, strong) NSString *currentText;


@end

@implementation YYRefreshView


#pragma mark - Initialization

- (instancetype)initWithPostion:(YYRefreshPosition)postion {
    if (self = [super init]) {
        self.postion = postion;
        [self setContext];
    }
    return self;
}


- (void)setContext {
    [self addSubview:self.textLabel];
    [self addSubview:self.imageView];
    [self addSubview:self.refreshingView];
}


#pragma mark - Override

- (void)layoutSubviews {
    [self updateLocation];
    [super layoutSubviews];
}

- (void)updateLocation {
    [_textLabel sizeToFit];
    CGPoint center = CGPointMake(CGRectGetMidX(self.bounds), CGRectGetMidY(self.bounds));
    _textLabel.center = center;
    
    center.x -= CGRectGetMidX(_textLabel.bounds) + CGRectGetMidX(_imageView.bounds) + 5;
    _imageView.center = center;
    _refreshingView.center = center;
}

#pragma mark - YYRefreshView

- (void)showWithState:(YYRefreshState)state config:(YYRefreshConfig *)config animated:(BOOL)animated {
    switch (state) {
        case YYRefreshStateIdle: {
            [self showIdleWithConfig:config animated:animated];
            break;
        }
        case YYRefreshStateReady: {
            [self showRedayWithConfig:config animated:animated];
            break;
        }
        case YYRefreshStateRefreshing: {
            [self showRefreshingWithConfig:config animated:animated];
            break;
        }
    }
}

#pragma mark - Private

- (void)showIdleWithConfig:(YYRefreshConfig *)config animated:(BOOL)animated {
    void (^action)(void) = ^{
        _imageView.transform = CGAffineTransformIdentity;
    };
    
    if (animated) {
        [UIView animateWithDuration:config.animationDurationSlow animations:action completion:nil];
    } else {
        action();
    }
    
    [_refreshingView stopAnimating];
    _imageView.hidden = NO;
    self.currentText = config.textIdle;
}

- (void)showRedayWithConfig:(YYRefreshConfig *)config animated:(BOOL)animated {
    
    void (^action)(void) = ^{
        _imageView.transform = CGAffineTransformRotate(_imageView.transform, DegreesToRadians(180.1));
    };
    
    if (animated) {
        [UIView animateWithDuration:config.animationDurationSlow animations:action completion:nil];
    } else {
        action();
    }
    self.currentText = config.textReady;
}

- (void)showRefreshingWithConfig:(YYRefreshConfig *)config animated:(BOOL)animated {
//    void (^action)(void) = ^{
//        _textLabel.text = config.textRefreshing;
//    };
//    
//    if (animated) {
//        [UIView animateWithDuration:YYRefreshSlowAnimationDuration animations:action completion:nil];
//    } else {
//        action();
//    }
    [_refreshingView startAnimating];
    _imageView.hidden = YES;
    self.currentText = config.textRefreshing;
}

#pragma mark - Setter

- (void)setCurrentText:(NSString *)currentText {
    _currentText = currentText;
    _textLabel.text = currentText;
    [_textLabel sizeToFit];
    [self layoutIfNeeded];
}

- (void)setPostion:(YYRefreshPosition)postion {
    if (_postion == postion) {
        return;
    }
    _postion = postion;
    switch (postion) {
        case YYRefreshPositionBottom:
        case YYRefreshPositionLeft:
            self.imageView.image = [UIImage imageNamed:YYRefreshImageUp];
            break;
        case YYRefreshPositionTop:
        case YYRefreshPositionRight:
            self.imageView.image = [UIImage imageNamed:YYRefreshImageDown];
            break;
    }
}

#pragma mark - Getter

- (UILabel *)textLabel {
    if (!_textLabel) {
        UILabel *label = [[UILabel alloc] init];
        label.font = YYRefreshLabelFont;
        label.textColor = YYRefreshLabelTextColor;
        label.textAlignment = NSTextAlignmentCenter;
        label.backgroundColor = [UIColor clearColor];
        label.text = YYRefreshIdleText;
        _textLabel = label;
    }
    return _textLabel;
}

- (UIImageView *)imageView {
    if (!_imageView) {
        UIImageView *imageView = [UIImageView new];
        imageView.bounds = CGRectMake(0, 0, 16, 16);
        _imageView = imageView;
    }
    return _imageView;
}

- (UIActivityIndicatorView *)refreshingView {
    if (!_refreshingView) {
        UIActivityIndicatorView *refreshingView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
        refreshingView.hidesWhenStopped = YES;
        _refreshingView = refreshingView;
    }
    return _refreshingView;
}


@end

File: YYRefreshSwift/YYRefreshSwift/YYRefresh/YYRefreshView.swift
//
//  YYRefreshView.swift
//  Swift3Sum
//
//  Created by sihuan on 2016/8/28.
//  Copyright © 2016年 huan. All rights reserved.
//

import UIKit

// MARK: - Public
extension YYRefreshView: YYRefreshViewProtocol {
    
    public func show(_ state: YYRefreshState, config: YYRefreshConfig, animated: Bool) {
        var action: (() -> Void)?
        var text: String?
        switch state {
        case .idle:
            action = {
                self.imageView.transform = CGAffineTransform.identity
            }
            refreshingView.stopAnimating()
            imageView.isHidden = false
            text = config.textIdle
        case .ready:
            action = {
                self.imageView.transform = CGAffineTransform(rotationAngle: degreesToRadians(180))
            }
            text = config.textReady
        case .refreshing:
            action = {
                self.imageView.transform = CGAffineTransform.identity
            }
            refreshingView.startAnimating()
            imageView.isHidden = true
            text = config.textRefreshing
        }
        
        setText(text)
        if let action = action {
            if animated {
                UIView.animate(withDuration: config.animationDurationSlow, animations: action)
            } else {
                action()
            }
        }
    }
}


public class YYRefreshView: UIView {
    
    // MARK: - Const
    let imageViewMargin: CGFloat = 5
    let imageViewWidth = 16
    let imageDown = "yy_arrow_down.png"
    let imageUp = "yy_arrow_up.png"
    
    // MARK: - Property
    var position: YYRefreshPosition
    
    lazy var imageView: UIImageView = {
        var imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: self.imageViewWidth, height: self.imageViewWidth))
        imageView.contentMode = .scaleAspectFit
        return imageView
    }()
    
    lazy var textLabel: UILabel = {
        var label = UILabel(frame: CGRect.zero)
        label.font = UIFont.boldSystemFont(ofSize: 14)
        label.textColor = UIColor(white: 0.4, alpha: 1.0)
        label.textAlignment = .center
        label.backgroundColor = UIColor.clear
        return label
    }()
    
    lazy var refreshingView: UIActivityIndicatorView = {
        var refreshingView = UIActivityIndicatorView(activityIndicatorStyle: .gray)
        refreshingView.hidesWhenStopped = true
        return refreshingView
    }()
    
    // MARK: - Initialization
    init(position: YYRefreshPosition) {
        self.position = position
        super.init(frame: CGRect.zero)
        setupUI()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func setupUI() {
        addSubview(textLabel)
        addSubview(imageView)
        addSubview(refreshingView)
        
        var imageName = ""
        switch position {
        case .left, .bottom:
            imageName = imageUp
        default:
            imageName = imageDown
        }
        
        imageView.image = imageFromBundle(named: imageName)
    }
}

// MARK: - Override
extension YYRefreshView {
    open override func layoutSubviews() {
        super.layoutSubviews()
        updateLocation()
    }
}

// MARK: - Private
extension YYRefreshView {
    func updateLocation() {
        var center = CGPoint(x: bounds.midX, y: bounds.midY)
        textLabel.sizeToFit()
        textLabel.center = center
        
        center.x -= textLabel.bounds.midX + imageView.bounds.midX + imageViewMargin
        imageView.center = center
        refreshingView.center = center
    }
    
    func setText(_ text: String?) {
        textLabel.text = text
        updateLocation()
    }
}


















